File: my_server/df703.py
import struct

class DF703:
    @staticmethod
    def byte_to_ascii_digits(byte):
        """Converts a single byte to two ASCII digits."""
        high_nibble = byte >> 4  # Extracts the high 4 bits
        low_nibble = byte & 0x0F  # Extracts the low 4 bits
        return f"{high_nibble}{low_nibble}"

    @staticmethod
    def parse_data_DF703(req_data):
        # Convert hex string to bytes if necessary
        if isinstance(req_data, str):
            req_data = bytes.fromhex(req_data)

        # Unpacking the common fields
        packet_head, forced_bit, device_type, report_data_type, packet_size, height, gps_selection = struct.unpack('>BBBBBHB', req_data[:8])

        parsed_data = {
            'Packet Head': f"{packet_head:02x}",
            'Forced Bit': f"{forced_bit:02x}",
            'Device Type': f"{device_type:02x}",
            'Report Data Type': f"{report_data_type:02x}",
            'Packet Size': f"{packet_size:02x}",
            'Height': height,
            'GPS Selection': f"{gps_selection:02x}"
        }

        # Conditional parsing based on GPS Selection
        offset = 8
        if gps_selection == 0x01:
            longitude, latitude = struct.unpack('>ff', req_data[offset:offset+8])
            offset += 8
            parsed_data.update({'Longitude': longitude, 'Latitude': latitude})
        
        temperature = req_data[offset]
        parsed_data['Temperature'] = temperature
        offset += 1

        # Parsing Reserved, Fall Angle, and Status
        reserved, fall_angle = struct.unpack('>BB', req_data[offset:offset+2])
        offset += 2
        parsed_data['Reserved'] = reserved
        parsed_data['Fall Angle'] = fall_angle

        # Status is two bytes
        status_bytes = req_data[offset:offset + 2]
        full_status = (status_bytes[0] & 0xF0) >> 4
        fire_status = status_bytes[0] & 0x0F
        fall_status = (status_bytes[1] & 0xF0) >> 4
        power_status = status_bytes[1] & 0x0F
        parsed_data.update({'Full Status': full_status, 'Fire Status': fire_status, 'Fall Status': fall_status, 'Power Status': power_status})
        offset += 2

        # Battery Voltage
        battery_voltage_hex = ''.join(f"{byte:02x}" for byte in req_data[offset:offset+2])
        battery_voltage_mV = int(battery_voltage_hex, 16) * 10
        parsed_data['Battery Voltage'] = battery_voltage_mV
        offset += 2

        # RSRP
        rsrp_bytes = req_data[offset:offset + 4]
        rsrp_float = struct.unpack('>f', rsrp_bytes)[0]
        parsed_data['RSRP'] = rsrp_float
        offset += 4

        # Frame Count
        frame_count_bytes = req_data[offset:offset + 2]
        frame_count_decimal = int.from_bytes(frame_count_bytes, byteorder='big')
        parsed_data['Frame Count'] = frame_count_decimal
        offset += 2

        # Device ID and IMEI
        device_id_bytes = req_data[offset:offset + 8]
        imei_bytes = device_id_bytes[1:]
        imei_str = ''.join(DF703.byte_to_ascii_digits(byte) for byte in imei_bytes)
        parsed_data['IMEI'] = imei_str

        return parsed_data

# Example usage
if __name__ == "__main__":
    data = b'\x80\x00\x01\x02\x1e\x04W\x00\x17\x00\x00\x00\x00\x01e\x00\x00x\xc4\x00\x01\x18e8P`\x06\x81Q\x81'
    parsed_data = DF703.parse_data(data)
    print(parsed_data)

File: Node-Red/node-red-contrib-df703/df703.js
module.exports = function(RED) {
    function DF703Node(config) {
        RED.nodes.createNode(this, config);

        this.on('input', function(msg) {
            const req_data = msg.payload;

            // Convert hex string to bytes if necessary
            if (typeof req_data === 'string') {
                req_data = Buffer.from(req_data, 'hex');
            }

            // Unpacking the common fields
            const packet_head = req_data.readUInt8(0);
            const forced_bit = req_data.readUInt8(1);
            const device_type = req_data.readUInt8(2);
            const report_data_type = req_data.readUInt8(3);
            const packet_size = req_data.readUInt8(4);
            const height = req_data.readUInt16BE(5);
            const gps_selection = req_data.readUInt8(7);

            const parsed_data = {
                'Packet Head': packet_head.toString(16).padStart(2, '0'),
                'Forced Bit': forced_bit.toString(16).padStart(2, '0'),
                'Device Type': device_type.toString(16).padStart(2, '0'),
                'Report Data Type': report_data_type.toString(16).padStart(2, '0'),
                'Packet Size': packet_size.toString(16).padStart(2, '0'),
                'Height': height,
                'GPS Selection': gps_selection.toString(16).padStart(2, '0'),
            };

            // Conditional parsing based on GPS Selection
            let offset = 8;
            if (gps_selection === 0x01) {
                const longitude = req_data.readFloatBE(offset);
                const latitude = req_data.readFloatBE(offset + 4);
                offset += 8;
                parsed_data['Longitude'] = longitude;
                parsed_data['Latitude'] = latitude;
            }

            const temperature = req_data.readUInt8(offset);
            parsed_data['Temperature'] = temperature;
            offset += 1;

            // Parsing Reserved, Fall Angle, and Status
            const reserved = req_data.readUInt8(offset);
            const fall_angle = req_data.readUInt8(offset + 1);
            offset += 2;
            parsed_data['Reserved'] = reserved;
            parsed_data['Fall Angle'] = fall_angle;

            // Status is two bytes
            const status_bytes = req_data.slice(offset, offset + 2);
            const full_status = (status_bytes[0] & 0xF0) >> 4;
            const fire_status = status_bytes[0] & 0x0F;
            const fall_status = (status_bytes[1] & 0xF0) >> 4;
            const power_status = status_bytes[1] & 0x0F;
            parsed_data['Full Status'] = full_status;
            parsed_data['Fire Status'] = fire_status;
            parsed_data['Fall Status'] = fall_status;
            parsed_data['Power Status'] = power_status;
            offset += 2;

            // Battery Voltage
            const battery_voltage_hex = req_data.slice(offset, offset + 2).toString('hex');
            const battery_voltage_mV = parseInt(battery_voltage_hex, 16) * 10;
            parsed_data['Battery Voltage'] = battery_voltage_mV;
            offset += 2;

            // RSRP
            const rsrp_bytes = req_data.slice(offset, offset + 4);
            const rsrp_float = rsrp_bytes.readFloatBE(0);
            parsed_data['RSRP'] = rsrp_float;
            offset += 4;

            // Frame Count
            const frame_count_bytes = req_data.slice(offset, offset + 2);
            const frame_count_decimal = frame_count_bytes.readUInt16BE(0);
            parsed_data['Frame Count'] = frame_count_decimal;
            offset += 2;

            // Device ID and IMEI
            const device_id_bytes = req_data.slice(offset, offset + 8);
            const imei_bytes = device_id_bytes.slice(1);
            const imei_str = Array.from(imei_bytes).map(byte => DF703.byte_to_ascii_digits(byte)).join('');
            parsed_data['IMEI'] = imei_str;

            msg.payload = parsed_data;
            this.send(msg);
        });
    }

    RED.nodes.registerType("df703", DF703Node);
}
